models/particle_net.py

```diff
@@ -0,0 +1,173 @@
+from omegaconf import OmegaConf, DictConfig
+import tensorflow as tf
+from tensorflow.keras.layers import BatchNormalization, Dense, Dropout, Conv2D, Activation
+
+
+class ParticleNet(tf.keras.Model):
+    def __init__(self, encoder_cfg, decoder_cfg):
+        super(ParticleNet, self).__init__()
+
+        self.num_classes = decoder_cfg['output_classes']
+        self.conv_params = [(layer_setup[0], tuple(layer_setup[1]),) for layer_setup in encoder_cfg['conv_params']]
+        self.conv_pooling = encoder_cfg['conv_pooling']
+        self.fc_params = [(layer_setup, encoder_cfg['dropout_rate'],) for layer_setup in encoder_cfg['dense_params']]
+        self.num_points = encoder_cfg['seq_cutoff_len']
+
+        # Initialising layers
+        self.batch_norm = BatchNormalization()
+        
+        self.edge_conv_layers = []
+        for layer_idx, layer_param in enumerate(self.conv_params):
+            K, channels = layer_param
+            self.edge_conv_layers.append(
+                EdgeConv(self.num_points, K, channels, with_bn=True, activation='relu', pooling=self.conv_pooling, name=f'{self.name}_EdgeConv_{layer_idx}')
+            )
+
+        self.dense_layers = []
+        self.dense_dropout = []
+        if self.fc_params is not None:
+            for layer_idx, layer_param in enumerate(self.fc_params):
+                units, dropout_rate = layer_param
+
+                self.dense_layers.append(
+                    Dense(units, activation='relu')
+                )
+
+                if dropout_rate is not None and dropout_rate > 0:
+                    self.dense_dropout.append(
+                        Dropout(dropout_rate)
+                    )
+            
+            self.out = Dense(self.num_classes, activation='softmax')
+
+    @tf.function
+    def call(self, input):
+        masking = True
+
+        features = input[0][:,:self.num_points,:].to_tensor() # (batch_size, particles, features), here: (128, 125, 22)
+        # print(f'FEATURES - type: {type(features)}, shape: {features.shape}')
+
+        # Converting from (r, theta)-space to (eta, phi)-space
```

Comment:
I don't know if I quite get what you mean. Do you mean that one can choose which type of coordinate system they want to use, i.e. choose between eta-phi or r-theta within the yaml?
If so, the distance function needs to be updated as it currently only works in eta-phi space.

Reply:
Yeah that would be nice to have it, but the way distance is computed is already hardwired into the model and it's hard to make it flexible straight away.

So my idea was to add a switch between using polar coordinates `["r", "theta"]` and user-defined set `["a", "b", "c", ...]`. It can be called `coordinate_type` and if it is `"polar"`, then there is assertion to be added `assert spatial_features == ["r", "theta"]` followed by current L46-52 without change. 

Else, it converts `spatial_features` to a list of indices, which are then gathered from the inputs into `points` similarly to how it is done in the embedding:
https://github.com/ofivite/tatoo/blob/6c1e019bb77f4fb2a0427a26e71c40a84603724b/models/embedding.py#L20